src/Stratis.Features.FederatedPeg/Wallet/MultiSigTransactions.cs

```diff
@@ -0,0 +1,185 @@
+
+        private bool TryAddWithdrawal(TransactionData transactionData)
+        {
+            uint256 matchingDepositId = transactionData.SpendingDetails?.WithdrawalDetails?.MatchingDepositId;
+
+            if (matchingDepositId == null)
+                return false;
+
+            if (!this.withdrawalsByDepositDict.TryGetValue(matchingDepositId, out List<TransactionData> txList))
+            {
+                txList = new List<TransactionData>();
+                this.withdrawalsByDepositDict.Add(transactionData.SpendingDetails.WithdrawalDetails.MatchingDepositId, txList);
+            }
+
+            txList.Add(transactionData);
+
+            return true;
+        }
+
+        private void TryRemoveWithdrawal(TransactionData transactionData)
+        {
+            uint256 matchingDepositId = transactionData.SpendingDetails?.WithdrawalDetails?.MatchingDepositId;
+
+            if (matchingDepositId == null)
+                return;
+
+            if (this.withdrawalsByDepositDict.TryGetValue(matchingDepositId, out List<TransactionData> txList))
+            {
+                txList.Remove(transactionData);
+
+                if (txList.Count == 0)
+                    this.withdrawalsByDepositDict.Remove(matchingDepositId);
+            }
+        }
+
+        public void Add(TransactionData transactionData)
+        {
+            lock (this.lockObject)
+            {
+                this.transactionDict.Add(transactionData.Key, transactionData);
+
+                if (transactionData.IsSpendable())
+                    this.spendableTransactionDict.Add(transactionData.Key, transactionData);
+
+                this.TryAddWithdrawal(transactionData);
+
+                transactionData.Subscribe(this);
+            }
+        }
+
+        public bool Remove(TransactionData transactionData)
+        {
+            lock (this.lockObject)
+            {
+                bool res = this.transactionDict.Remove(transactionData.Key);
+
+                if (this.spendableTransactionDict.ContainsKey(transactionData.Key))
+                    this.spendableTransactionDict.Remove(transactionData.Key);
+
+                this.TryRemoveWithdrawal(transactionData);
+
+                return res;
+            }
+        }
+
+        public IEnumerator<TransactionData> GetEnumerator()
+        {
+            lock (this.lockObject)
+            {
+                return this.transactionDict.Values.ToList().GetEnumerator();
+            }
+        }
+
+        IEnumerator IEnumerable.GetEnumerator()
+        {
+            return this.transactionDict.Values.ToList().GetEnumerator();
+        }
+
+        public void Clear()
+        {
+            lock (this.lockObject)
+            {
+                this.transactionDict.Clear();
+                this.spendableTransactionDict.Clear();
+                this.withdrawalsByDepositDict.Clear();
+            }
+        }
+
+        public void BeforeSpendingDetailsChanged(TransactionData transactionData)
```

Comment:
I'm not a fan of naming methods "where they are being called" rather than "what they do"

Reply:
It's up to the observer's method implementation to decide what to do but it makes for better decoupling if the interface, and corresponding method naming, makes no assumption as to that implementation.